python/ray/autoscaler/_private/kuberay/node_provider.py

```diff
@@ -0,0 +1,295 @@
+        for index, spec in enumerate(worker_group_specs):
+            if spec["groupName"] == group_name:
+                group_index = index
+                group_spec = spec
+                break
+        assert group_index is not None and group_spec is not None
+        return group_index, group_spec
+
+    def _wait_for_pods(self, group_name: str, replicas: int) -> None:
+        """Wait until `replicas` pods of type group_name are posted."""
+        label_filters = to_label_selector({
+            "ray.io/cluster": self.cluster_name,
+            "ray.io/group": group_name
+        })
+        while True:
+            pods = self._get("pods?labelSelector=" +
+                             requests.utils.quote(label_filters))
+            logger.info(
+                "Currently have {} replicas of group {}, requested {}.".format(
+                    len(pods["items"]), group_name, replicas))
+            if len(pods["items"]) == replicas:
+                break
+            else:
+                logger.info(
+                    "Waiting for reconciler, number of replicas is {}, expected {}.".
+                    format(len(pods["items"]), replicas))
+                time.sleep(10.0)
+
+    def create_node(self, node_config: Dict[str, Any], tags: Dict[str, str],
+                    count: int) -> Dict[str, Dict[str, str]]:
+        """Creates a number of nodes within the namespace."""
+        with self._lock:
+            url = "rayclusters/{}".format(self.cluster_name)
+            raycluster = self._get(url)
+            group_name = tags["ray-user-node-type"]
+            group_index, group_spec = self._get_worker_group(
+                raycluster, group_name)
+            path = f"/spec/workerGroupSpecs/{group_index}/replicas"
+            payload = [{
+                "op": "test",
+                "path": path,
+                "value": group_spec["replicas"]
+            }, {
+                "op": "replace",
+                "path": path,
+                "value": group_spec["replicas"] + count
+            }]
+            self._patch(url, payload)
+            self._wait_for_pods(group_name, group_spec["replicas"] + count)
+        return {}
+
+    def internal_ip(self, node_id: str) -> str:
+        """Get internal IP of a node (= Kubernetes pod)."""
+        data = self._get("pods/{}".format(node_id))
+        return data["status"].get("podIP", "IP not yet assigned")
+
+    def node_tags(self, node_id: str) -> Dict[str, str]:
+        """Get tags of a node (= Kubernetes pod)."""
+        data = self._get("pods/{}".format(node_id))
+        return make_node_tags(data["metadata"]["labels"], status_tag(data))
+
+    def non_terminated_nodes(self, tag_filters: Dict[str, str]) -> List[str]:
+        """Return a list of node ids filtered by the specified tags dict."""
+        data = self._get("pods/")
+
+        result = []
+        for pod in data["items"]:
+            labels = pod["metadata"]["labels"]
```

Comment:
Thanks, I changed it so the cluster is filtered with label selectors, but the labels unfortunately can't be in general (since the they need to be converged to tags, which needs the status tag). Filtering on the cluster gives us all of the efficiency benefit that is to be gained by filters though, so we are good here :)

Reply:
Ohh, right -- we insert the status tag. Looks good.